File: leetcode/0041-first-missing-positive.ts
/**
 * Finds the first missing positive integer in an array
 * Uses cyclic sort to place each number at its correct index position
 * Time complexity: O(n), Space complexity: O(1)
 * 
 * @param nums - Array of integers
 * @returns The smallest missing positive integer
 */
function firstMissingPositive(nums: number[]): number {
    const arrayLength: number = nums.length;
  
    // Step 1: Place each positive number at its correct index position
    // Number 1 should be at index 0, number 2 at index 1, etc.
    for (let currentIndex = 0; currentIndex < arrayLength; currentIndex++) {
        // Keep swapping while:
        // - Current number is in valid range [1, n]
        // - Current number is not already at its correct position
        while (nums[currentIndex] >= 1 && 
               nums[currentIndex] <= arrayLength && 
               nums[currentIndex] !== nums[nums[currentIndex] - 1]) {
          
            // Calculate target index for the current number
            const targetIndex: number = nums[currentIndex] - 1;
          
            // Swap current number with the number at its target position
            [nums[currentIndex], nums[targetIndex]] = [nums[targetIndex], nums[currentIndex]];
        }
    }
  
    // Step 2: Find the first position where number doesn't match (index + 1)
    for (let index = 0; index < arrayLength; index++) {
        // If number at index i is not (i + 1), then (i + 1) is missing
        if (nums[index] !== index + 1) {
            return index + 1;
        }
    }
  
    // Step 3: All numbers from 1 to n are present, so return n + 1
    return arrayLength + 1;
}

File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
    /**
     * Finds the smallest missing positive integer in an unsorted array.
     * Uses cyclic sort to place each positive number at its corresponding index.
     * Time Complexity: O(n), Space Complexity: O(1)
     * 
     * @param nums the input array
     * @return the smallest missing positive integer
     */
    public int firstMissingPositive(int[] nums) {
        int arrayLength = nums.length;
      
        // Phase 1: Place each positive number at its correct position
        // For a number k where 1 <= k <= n, place it at index k-1
        for (int currentIndex = 0; currentIndex < arrayLength; currentIndex++) {
            // Keep swapping while:
            // 1. Current number is positive
            // 2. Current number is within valid range [1, n]
            // 3. Current number is not already at its correct position
            while (nums[currentIndex] > 0 && 
                   nums[currentIndex] <= arrayLength && 
                   nums[currentIndex] != nums[nums[currentIndex] - 1]) {
                // Swap current number to its correct position
                int targetIndex = nums[currentIndex] - 1;
                swap(nums, currentIndex, targetIndex);
            }
        }
      
        // Phase 2: Find the first position where number doesn't match index + 1
        for (int index = 0; index < arrayLength; index++) {
            if (nums[index] != index + 1) {
                // The missing positive is index + 1
                return index + 1;
            }
        }
      
        // All numbers from 1 to n are present, so the answer is n + 1
        return arrayLength + 1;
    }
  
    /**
     * Helper method to swap two elements in the array
     * 
     * @param nums the array
     * @param firstIndex index of the first element
     * @param secondIndex index of the second element
     */
    private void swap(int[] nums, int firstIndex, int secondIndex) {
        int temp = nums[firstIndex];
        nums[firstIndex] = nums[secondIndex];
        nums[secondIndex] = temp;
    }
}
